### PR TITLE
fix the function find_common_types bug

### PR DESCRIPTION
**Pull request by ghost**
` types[0]` can raise a KeyError when `types` is a `pd.Series` . see issue #25270

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Review by WillAyd — CHANGES_REQUESTED** [review]
Haven't reviewed all of the failures but this doesn't seem right given this is a very generic function. Does the error affect things outside of SparseDataFrame? If not then seems like the issue needs to be addressed directly there

**Comment by WillAyd** [comment]
Also please add test(s) - should be the first part to any PR

**Comment by yichen8** [comment]
Ee, how to add test(s)? :smile:

File: pandas/core/dtypes/cast.py [hunk 1]
```diff
@@ -1075,7 +1075,7 @@ def find_common_type(types):
 
     Parameters
     ----------
-    types : list of dtypes
+    types :  list_like
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: list-like

File: pandas/core/dtypes/cast.py [hunk 2]
```diff
@@ -1090,7 +1090,7 @@ def find_common_type(types):
     if len(types) == 0:
         raise ValueError('no types given')
 
-    first = types[0]
+    first = types[:1]
```
Comment (jreback) [inline pandas/core/dtypes/cast.py]: if you are changing ths, you must have a failing test case, can you pls add it

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by yichen8** [comment]
In issues #25270 @rasbt gave  this queston. He thinks that "the Pandas SparseDataFrame method `to_coo()` (and possibly others) cannot handle sparse dataframes if the column names are integer types and don't start at 0. If the column names start at 0 or are string types, this is not an issue." 

Yes, he is right. Then I try to track the `KeyError` and find a wrong in the function `find_common_type()` . I will show how this error occur.

```python
# example from @rasbt

import pandas as pd
import numpy as np

ary = np.array([ [1, 0, 0, 3],
                 [1, 0, 2, 0],
                 [0, 4, 0 ,0] ])

df = pd.DataFrame(ary)
df.columns = [1, 2, 3, 4]

dfs = pd.SparseDataFrame(df,
                         default_fill_value=0)

# DOES NOT WORK:

dfs.to_coo() # raises KeyError: 0
```

now if we check:

```python
In [12]: dfs.dtypes
Out[12]: 
1    int64
2    int64
3    int64
4    int64
dtype: object
In [13]: type(dfs.dtypes)
Out[13]: pandas.core.series.Series    
```

as we see, the `dfs.dtype` is not a `list` , and work calls the function `find_comm_type()` :

```python
# pandas/core/dtypes/cast.py in find_common_type(types) at about 1093 lines
def find_common_type(types):
    """
    Find a common data type among the given dtypes.

    Parameters
    ----------
    types : list of dtypes

    Returns
    -------
    pandas extension or numpy dtype

    See Also
    --------
    numpy.find_common_type

    """

    if len(types) == 0:
        raise ValueError('no types given')

    first = types[0] # list is ok, but pd.Series may cause litte error.
```

We check this statement `first = types[0]`:

```python
In [20]: dfs.dtypes[0]
---------------------------------------------------------------
KeyError                      Traceback (most recent call last)
<ipython-input-20-4d14dd9f5c73> in <module>()
----> 1 dfs.dtypes[0]

~/anaconda3/lib/python3.7/site-packages/pandas/core/series.py in __getitem__(self, key)
    765         key = com._apply_if_callable(key, self)
    766         try:
--> 767             result = self.index.get_value(self, key)
    768 
    769             if not is_scalar(result):

~/anaconda3/lib/python3.7/site-packages/pandas/core/indexes/base.py in get_value(self, series, key)
   3116         try:
   3117             return self._engine.get_value(s, k,
-> 3118                                           tz=getattr(series.dtype, 'tz', None))
   3119         except KeyError as e1:
   3120             if len(self) > 0 and self.inferred_type in ['integer', 'boolean']:

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_value()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_value()

pandas/_libs/index.pyx in pandas._libs.index.IndexEngine.get_loc()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.Int64HashTable.get_item()

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.Int64HashTable.get_item()

KeyError: 0
```

Yes, it raises a `KeyError`.  So using `types[0]` to get the `types` first item can't pass in this case.

Then in the example of @rasbt there are two cases can work.

```python
# WORKS (1)

dfs2 = dfs.copy()
dfs2.columns = [0, 1, 2, 3]
dfs2.to_coo()

# WORKS (2)

dfs3 = dfs.copy()
dfs3.columns = [str(i) for i in dfs3.columns]
dfs3.to_coo()
```

In fact, `dfs.columns` will be `dfs.dtypes.index`.  Now, `dfs.dtypes` and `dfs2.dtypes`,`dfs3.dtypes`are `Series`.  And We know `Series` has some **features**.

```python
In [10]: dfs.dtypes.index
Out[10]: Int64Index([1, 2, 3, 4], dtype='int64')

In [11]: dfs2.dtypes.index
Out[11]: Int64Index([0, 1, 2, 3], dtype='int64')

In [12]: dfs3.dtypes.index
Out[12]: Index(['1', '2', '3', '4'], dtype='object')
```

Useing `types[0]` will take different ways for different `Series.index.dtype`.  (Of couse, is `pd.DataFrame` too.) `0` in `types[0]` is regard as a key when `dfs.dtypes.index.dtype` is int64, but a index when `dfs.dtypes.index.dtype` is object (str) like  `list[0]`. 

So `first = types[0]` can't handle those.  and `first = [t for t in types][0]` will solve those simply.

Of cause, `first = types[:1]` is not a right way , beacuse it can't pass test although it can slove @rasbt's  case.

But after committing, some checks were not successful, the newest update passed test.

**Comment by jreback** [comment]
closing as stale if you want to keep working, merge master and ping

**Pull request closed without merging**